src/main/java/io/vertx/core/impl/utils/CyclicSequence.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2011-2017 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.impl.utils;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A thread safe cyclic sequence of elements that can be used for round robin.
+ * <p/>
+ * A sequence is immutable and mutations uses {@link #add(Object)} and {@link #remove(Object)}
+ * to return a modified copy of the current instance.
+ * <p/>
+ * The iterator uses a volatile index, so it can be incremented concurrently by several
+ * threads with locking.
+ *
+ * @author <a href="http://tfox.org">Tim Fox</a>
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+public class CyclicSequence<T> implements Iterable<T>, Iterator<T> {
+
+  private final AtomicInteger pos;
+  private final Object[] elements;
```

Comment:
I am still convinced that using an array instead of list makes the code more complex without significant gain in return.

Reply:
we use arrays because we can make them final and be sure the array is published correctly with respect to the java memory model. An ArrayList does not have its internal array final (also it removes one extra indirection). That's what CopyOnWriteArrayList does as well internally.

Reply:
so the gain is to have safe publication

Reply:
Where is this field published? In your previous implementation, `Handlers` only provided an Iterator for the content. That is not publishing the array itself.

Reply:
it's published in the JMM meaning (https://shipilev.net/blog/2014/safe-public-construction/). Previously Handlers was using a CopyOnWriteArrayList that has it's internal array to be volatile. In this case we are using a final value.

Reply:
the CyclicList is copied and then put in the ConcurrentHashMap, we use final to ensure that the copied array is published in CyclicList that is then made available to the concurrent hashmap via merge/compute operations